File: src/main/java/nz/co/searchwellington/controllers/ajax/BaseAjaxController.java
package nz.co.searchwellington.controllers.ajax;

import java.io.IOException;
import java.util.Collections;
import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.web.servlet.ModelAndView;

import uk.co.eelpieconsulting.common.views.ViewFactory;

public abstract class BaseAjaxController {
	    
    private static final String TERM = "term";
    
    protected ViewFactory viewFactory;

	public ModelAndView handleRequest(HttpServletRequest request, HttpServletResponse response) throws IOException {
		final ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        List<String> suggestions = Collections.emptyList();
        if (request.getParameter(TERM) != null) {
        	suggestions = this.getSuggestions(request.getParameter(TERM));
        }
        mv.addObject("data", suggestions);
        return mv;
    }
    
    protected abstract List<String> getSuggestions(String q);	
    
}

File: src/main/scala/nz/co/searchwellington/controllers/ajax/BaseAjaxController.scala
package nz.co.searchwellington.controllers.ajax

import jakarta.servlet.http.{HttpServletRequest, HttpServletResponse}
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.LoggedInUserFilter
import nz.co.searchwellington.model.User
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.common.views.ViewFactory

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}
import scala.jdk.CollectionConverters._

// TODO should be able to dog food these end points by consuming the json put of the main publishers and tags models
abstract class BaseAjaxController extends ReasonableWaits {

  private val TERM = "term"

  def loggedInUserFilter: LoggedInUserFilter

  def viewFactory: ViewFactory

  def handleRequest(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    val eventualSuggestions = Option(request.getParameter(TERM)).map { q =>
      getSuggestions(q, loggedInUserFilter.getLoggedInUser)
    }.getOrElse {
      Future.successful(Seq.empty)
    }

    Await.result(for {
      suggestions <- eventualSuggestions
    } yield {
      val mv = new ModelAndView(viewFactory.getJsonView)
      mv.addObject("data", suggestions.asJava)
      mv
    }, TenSeconds)
  }

  protected def getSuggestions(q: String, loggedInUser: Option[User]): Future[Seq[String]]

}
